File: document/documentation/reference/operation/routing/d8_flow_direction/example-1/d8_flow_direction.cpp
#include "lue/framework/api/cxx.hpp"
#include "lue/document.hpp"
#include <hpx/hpx_main.hpp>


class Example: public lue::document::Example
{

    public:

        Example(int const argc, char const* const* argv):

            lue::document::Example{
                [&]() -> cxxopts::Options
                {
                    cxxopts::Options options{std::filesystem::path{argv[0]}.filename()};

                    options.positional_help("<elevation> <flow_direction>");
                    options.add_options()
                        // clang-format off
                        ("elevation_array", "array to read", cxxopts::value<std::string>())
                        ("flow_direction_array", "array to write", cxxopts::value<std::string>())
                        // clang-format on
                        ;
                    options.parse_positional({"elevation_array", "flow_direction_array"});

                    return options;
                }(),
                argc,
                argv}

        {
        }

    private:

        auto run_implementation() -> int override
        {
            if (!argument_parsed("elevation_array") || !argument_parsed("flow_direction_array"))
            {
                throw std::runtime_error("Wrong usage, pass '--help' for more information");
            }

            auto const elevation_array_pathname = argument<std::string>("elevation_array");
            auto const flow_direction_array_pathname = argument<std::string>("flow_direction_array");

            using namespace lue::api;

            // [example
            Field const elevation = from_gdal(elevation_array_pathname);
            Field const flow_direction = d8_flow_direction(elevation);

            auto written = to_gdal(flow_direction, flow_direction_array_pathname);
            // example]

            written.get();
            lue::document::write_value_scale(flow_direction_array_pathname, "flow_direction");

            return EXIT_SUCCESS;
        }
};


auto main(int argc, char** argv) -> int
{
    return Example{argc, argv}.run();
}

File: document/documentation/reference/operation/routing/d8_flow_direction/example-1/d8_flow_direction.py
import sys
from pathlib import Path

import docopt
import lue.framework as lfrx
from lue import __version__ as version


@lfrx.runtime_scope
def main() -> None:
    command = Path(sys.argv[0]).name
    usage = f"""\
Usage:
    {command} <elevation> <flow_direction>

Options:
    -h --help         Show this screen and exit
    --version         Show version and exit
    <elevation>       Elevation array to read
    <flow_direction>  Flow direction array to write
"""
    arguments = docopt.docopt(usage, sys.argv[1:], version=version)
    elevation_array_path = Path(arguments["<elevation>"])
    flow_direction_array_path = Path(arguments["<flow_direction>"])

    # [example
    elevation = lfrx.from_gdal(str(elevation_array_path))
    flow_direction = lfrx.d8_flow_direction(elevation)

    lfrx.to_gdal(flow_direction, str(flow_direction_array_path))
    # example]


if __name__ == "__main__":
    sys.exit(main())
